src/updates.c

```diff
@@ -0,0 +1,706 @@
+
+  return TRUE;
+}
+
+typedef struct _UpdatesSession
+{
+  Session parent;
+  char *ref;
+  FlatpakInstallation *installation;
+  guint timeout;
+
+  gboolean closed;
+} UpdatesSession;
+
+typedef struct _UpdatesSessionClass
+{
+  SessionClass parent_class;
+} UpdatesSessionClass;
+
+GType updates_session_get_type (void);
+
+G_DEFINE_TYPE (UpdatesSession, updates_session, session_get_type ())
+
+static void
+updates_session_close (Session *session)
+{
+  UpdatesSession *updates_session = (UpdatesSession *)session;
+
+  updates_session->closed = TRUE;
+
+  g_debug ("updates session owned by '%s' closed", session->sender);
+}
+
+static void
+updates_session_finalize (GObject *object)
+{
+  UpdatesSession *session = (UpdatesSession*)object;
+
+  if (session->timeout)
+    g_source_remove (session->timeout);
+  g_free (session->ref);
+  g_clear_object (&session->installation);
+
+  G_OBJECT_CLASS (updates_session_parent_class)->finalize (object);
+}
+
+static void
+updates_session_init (UpdatesSession *session)
+{
+}
+
+static void
+updates_session_class_init (UpdatesSessionClass *klass)
+{
+  GObjectClass *object_class;
+  SessionClass *session_class;
+
+  object_class = G_OBJECT_CLASS (klass);
+  object_class->finalize = updates_session_finalize;
+
+  session_class = (SessionClass *)klass;
+  session_class->close = updates_session_close;
+}
+
+static gboolean
+check_for_updates (gpointer data)
+{
+  Session *session = data;
+  UpdatesSession *usession = data;
+  g_autoptr(GPtrArray) updates = NULL;
+  int i;
+
+  g_debug ("Checking for updates for %s", usession->ref);
+
+  updates = flatpak_installation_list_installed_refs_for_update (usession->installation, NULL, NULL);
+
+  for (i = 0; i < updates->len; i++)
+    {
+      FlatpakInstalledRef *iref = g_ptr_array_index (updates, i);
+      g_autofree char *ref = flatpak_ref_format_ref (FLATPAK_REF (iref));
+      if (g_str_equal (ref, usession->ref))
+        {
+          GVariantBuilder builder;
+          g_autoptr(GError) error = NULL;
+
+          g_debug ("Found update for %s", usession->ref);
+
+          g_variant_builder_init (&builder, G_VARIANT_TYPE_VARDICT);
+
+          if (!g_dbus_connection_emit_signal (session->connection,
+                                              session->sender,
+                                              "/org/freedesktop/portal/desktop",
+                                              "org.freedesktop.portal.Updates",
+                                              "UpdateAvailable",
+                                              g_variant_new ("(oa{sv})", session->id, &builder),
+                                              &error))
+            {
+              g_warning ("Failed to emit UpdateAvailable: %s", error->message);
+            }
+          break;
+        }
+    }
+
+  return G_SOURCE_CONTINUE;
+}
+
+static void
+start_monitoring (UpdatesSession *session)
+{
+  check_for_updates (session);
```

Comment:
Maybe we could have a shared poll, rather than one per monitor? Because the poll is going to do some network i/o.